x/concentrated-liquidity/swaps_test.go

```diff
@@ -1361,3 +1125,164 @@ func (s *KeeperTestSuite) TestSwapExactAmountOut() {
+
+			s.Require().Equal(poolBefore.GetTotalShares(), poolAfter.GetTotalShares())
+			s.Require().Equal(poolBefore.GetTotalPoolLiquidity(s.Ctx), poolAfter.GetTotalPoolLiquidity(s.Ctx))
+			s.Require().Equal(oldSpotPrice, newSpotPrice)
+
+			// 3. assure that user balacne came back to original
+			userBalanceAfterSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, s.TestAccs[0])
+			poolBalanceAfterSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, poolBefore.GetAddress())
+
+			s.Require().Equal(userBalanceBeforeSwap, userBalanceAfterSwap)
+			s.Require().Equal(poolBalanceBeforeSwap, poolBalanceAfterSwap)
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestInverseRelationshipSwapInAmtGivenOut() {
+	tests := swapInGivenOutCases
+
+	for name, test := range tests {
+		s.Run(name, func() {
+			s.Setup()
+			s.FundAcc(s.TestAccs[0], sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(10000000000000)), sdk.NewCoin("usdc", sdk.NewInt(1000000000000))))
+			s.FundAcc(s.TestAccs[1], sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(10000000000000)), sdk.NewCoin("usdc", sdk.NewInt(1000000000000))))
+
+			// Create default CL pool
+			pool := s.PrepareConcentratedPool()
+
+			// add default position
+			_, _, _, err := s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, pool.GetId(), s.TestAccs[0], DefaultAmt0, DefaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), DefaultLowerTick, DefaultUpperTick)
+			s.Require().NoError(err)
+
+			// add second position depending on the test
+			if !test.secondPositionLowerPrice.IsNil() {
+				newLowerTick, err := math.PriceToTick(test.secondPositionLowerPrice, DefaultExponentAtPriceOne) // 315000
+				s.Require().NoError(err)
+				newUpperTick, err := math.PriceToTick(test.secondPositionUpperPrice, DefaultExponentAtPriceOne) // 322500
+				s.Require().NoError(err)
+
+				_, _, _, err = s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, pool.GetId(), s.TestAccs[1], DefaultAmt0, DefaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), newLowerTick.Int64(), newUpperTick.Int64())
+				s.Require().NoError(err)
+			}
+
+			// mark pool state and user balance before swap
+			poolBefore, err := s.App.ConcentratedLiquidityKeeper.GetPool(s.Ctx, pool.GetId())
+			s.Require().NoError(err)
+			userBalanceBeforeSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, s.TestAccs[0])
+			poolBalanceBeforeSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, poolBefore.GetAddress())
+
+			// system under test
+			fmt.Println("======")
+			firstTokenIn, firstTokenOut, _, _, sqrtPrice, err := s.App.ConcentratedLiquidityKeeper.SwapInAmtGivenOut(
+				s.Ctx,
+				test.tokenOut, test.tokenInDenom,
+				DefaultZeroSwapFee, test.priceLimit, pool.GetId())
+			fmt.Printf("first swap token in %s\n", firstTokenIn.String())
+			fmt.Printf("first swap token out %s\n", firstTokenOut.String())
+
+			secondTokenIn, secondTokenOut, _, _, _, err := s.App.ConcentratedLiquidityKeeper.SwapOutAmtGivenIn(
+				s.Ctx,
+				firstTokenOut, firstTokenIn.Denom,
+				DefaultZeroSwapFee, sqrtPrice.Power(2), pool.GetId(),
```

Comment:
hey @czarcas7ic do you have suggestion of what I could be using as the value for price limit here? This is the only part of this PR that remains unfinished.

Bascially, here we're using the testing vectors you've created, take the results and then use them again for swapping. I',m currently stuck on choosing which value to use for the price limit for the second swap here 🤔

Reply:
Sorry just now catching up on notifications, I believe I answered this on the call with examples. Please lmk if you need further examples!